app/controllers/users/registrations_controller.rb

```diff
@@ -1,10 +1,41 @@
-# frozen_string_literal: true
-
 class Users::RegistrationsController < Devise::RegistrationsController
+  alias :rebuild_devise_user :build_resource
+
+  def create
+    if email_matches_whitelist
+      do_default_devise_behaviour
+    else
+      rebuild_devise_user
+      add_error_to_devise_user
+      render 'new'
+    end
+  end
+
 protected
 
   # The path used after sign up for inactive accounts.
   def after_inactive_sign_up_path_for(_resource)
     users_confirmations_pending_path
   end
+
+  def do_default_devise_behaviour
+    Devise::RegistrationsController
+      .instance_method(:create).bind(self).call
+  end
+
+  def email_matches_whitelist
```

Comment:
Wonder if we can evolve this repo to also have Actors?  Be nice to see this behaviour separated from the Devise fun.

Reply:
I considered putting this in use case, but wasn't sure if it was worth it - definitely planning to pull it out when whitelist evolves into a more complex test (or regex, ugh).

Could pull it out now if you feel strongly about it - might help to give some shape to the architecture early on.